src/containers/containers.c

```diff
@@ -289,6 +293,340 @@ extern inline container_t *container_andnot(
+    }
+}
+
+uint32_t container_iterator_read_into_uint32(const container_t *c,
+                                             uint8_t typecode,
+                                             roaring_container_iterator_t *it,
+                                             uint32_t high16, uint32_t *buf,
+                                             uint32_t count) {
+    uint32_t consumed = 0;
+    switch (typecode) {
+        case BITSET_CONTAINER_TYPE: {
+            const bitset_container_t *bc = const_CAST_bitset(c);
+            uint32_t wordindex = it->in_container_index / 64;
+            uint64_t word = bc->words[wordindex] &
+                            (UINT64_MAX << (it->in_container_index % 64));
+            do {
+                // Read set bits.
+                while (word != 0 && consumed < count) {
+                    *buf = high16 |
+                           (wordindex * 64 + roaring_trailing_zeroes(word));
+                    word = word & (word - 1);
+                    buf++;
+                    consumed++;
+                }
+                // Skip unset bits.
+                while (word == 0 &&
+                       wordindex + 1 < BITSET_CONTAINER_SIZE_IN_WORDS) {
+                    wordindex++;
+                    word = bc->words[wordindex];
+                }
+            } while (word != 0 && consumed < count);
+
+            it->has_value = word != 0;
+            if (it->has_value) {
+                it->in_container_index =
+                    wordindex * 64 + roaring_trailing_zeroes(word);
+                it->value = it->in_container_index;
+            }
+            return consumed;
+        }
+        case ARRAY_CONTAINER_TYPE: {
+            const array_container_t *ac = const_CAST_array(c);
+            uint32_t num_values =
+                minimum_uint32(ac->cardinality - it->in_container_index, count);
+            for (uint32_t i = 0; i < num_values; i++) {
+                buf[i] = high16 | ac->array[it->in_container_index + i];
+            }
+            buf += num_values;
+            consumed += num_values;
+            it->in_container_index += num_values;
+            it->has_value = it->in_container_index < ac->cardinality;
+            if (it->has_value) {
+                it->value = ac->array[it->in_container_index];
+            }
+            return consumed;
+        }
+        case RUN_CONTAINER_TYPE: {
+            const run_container_t *rc = const_CAST_run(c);
+            do {
+                uint32_t largest_run_value = rc->runs[it->run_index].value +
+                                             rc->runs[it->run_index].length;
+                uint32_t num_values = minimum_uint32(
+                    largest_run_value - it->value + 1, count - consumed);
+                for (uint32_t i = 0; i < num_values; i++) {
+                    buf[i] = high16 | (it->value + i);
+                }
+                it->value += num_values;  // TODO: can this overflow?
```

Comment:
Yes, I believe this can overflow. Consider this case:

The iterator currently is at `has_value = true`, `value = UINT16_MAX`, and you have a run container with `value = UINT16_MAX`, `length = 0`, and asking for `count = 10`

1. largest_run_value = `UINT16_MAX`
2. num_values = `min(UINT16_MAX - UINT16_MAX + 1, 10 - 0)` = `1`
3. it->value = `UINT16_MAX + 1` = `0`

I think this would be worth a comment below where we check `it->value == 0`